### DISCUSSION TITLE
new plugin hooks

### DISCUSSION THREAD

**Discussion by pathardy** [fudforum/FUDforum#33] [Ideas]
In the documentation wiki there is a set of plugin hooks currently provided and a note that further hooks can be added.

I propose adding four hooks. Below is a brief description of each. Note that three pass an array of arguments with named indices; if there is a preferred/canonical way to pass multiple arguments to plugin_call_hook(), please advise. Let me know what else (patches?) you need, as well as any comments or questions.

'GROUP_JOIN'

This hook is placed in groupmgr.php.t just after adding a new record to the group_members table.

plugin_call_hook('GROUP_JOIN', array('uid'=>$usr_id, 'gid'=>$group_id));

'GROUP_LEAVE'

This hook is placed in grp_delete_member() in groups.inc.t.

plugin_call_hook('GROUP_LEAVE', array('uid'=>$user_id, 'gid'=>$id));

'CHANGE_OF_ADDRESS'

This hook is placed in register.php.t and executed when a user changes their email address.

plugin_call_hook('CHANGE_OF_ADDRESS', array('uid'=>$uid, 'old_email'=>$old_email, 'new_email'=>$uent->email));
  

'ACCOUNT_DELETE'

This hook is placed in admuser.php after an administrator has confirmed the deletion of an account, but just before the call to usr_delete().

plugin_call_hook('ACCOUNT_DELETE', $usr_id) ;

**Comment by naudefj** [comment]
> if there is a preferred/canonical way to pass multiple arguments to plugin_call_hook(), please advise

Your code looks valid. Here is an example from iemail.inc.t:
   list($to, $subj, $body, $header) = plugin_call_hook('PRE_MAIL', array($to, $subj, $body, $header));

**Comment by naudefj** [comment]
> Let me know what else (patches?) you need

A patch would help a lot!
Otherwise, just give an indication of where the call-outs should go (line number, before or after code, etc.) and I will try to add them for you.

**Reply by pathardy** [reply]
[proposed-hooks.tar.gz](https://github.com/fudforum/FUDforum/files/11494507/proposed-hooks.tar.gz)

**Comment by naudefj** [comment]
Patch committed as per https://github.com/fudforum/FUDforum/commit/3c85826a3b8fda405b413f9e2f52ecef3e2e50d5

Note that I've changed 'CHANGE_OF_ADDRESS' to ''EMAIL_CHANGED'. 
Hope it won't cause any problems.

PS: Documentation also updated @ https://github.com/fudforum/FUDforum/wiki/Plugins
